Lib/importlib/_bootstrap_external.py

```diff
@@ -91,10 +92,33 @@ def _unpack_uint16(data):
     return int.from_bytes(data, 'little')
 
 
-def _path_join(*path_parts):
-    """Replacement for os.path.join()."""
-    return path_sep.join([part.rstrip(path_separators)
-                          for part in path_parts if part])
+if _MS_WINDOWS:
+    def _path_join(*path_parts):
+        """Replacement for os.path.join()."""
```

Comment:
This implementation of `_path_join` doesn't support "\\\\?\\" paths. But I haven't checked how this plays out in practice with the use of `_path_join` in `FileFinder` when the script directory in `sys.path[0]` is a "\\\\?\\" path, based on the path that's used to run the script.

I'd rather have a `_norm_splitdrive` function here. For example, "//?/////unc///server///share///" (valid path) would be normalized and split as `('\\\\?\\UNC\\server\\share', '\\')`. Normalizing makes splitting the drive easier since repeated separators can be collapsed. That should be fine here since it's for internal use. Once you have the drive split out, the process of joining components is at a higher level. For the most part, you don't have to think about the contents of the drive. For example:

```python
def _path_join(*path_parts):
    sep = '\\'
    drive = path = ''
    for part in path_parts:
        if not part:
            continue
        p_drive, p_path = _norm_splitdrive(part)
        if p_drive and p_drive.upper() != drive.upper():
            drive = p_drive
            path = p_path
        elif not path or p_path[:1] == sep:
            if drive[:1] != sep:
                path = p_path
            else:
                path = f'\\{p_path.lstrip(sep)}'
        elif p_path:
            path = f'{path.rstrip(sep)}\\{p_path}'
    return f'{drive}{path}'
```
The major problem with a splitdrive-based implementation is performance.

Reply:
No, the major problem with this is that we've got to maintain two implementations of the same complex algorithm :)

I'm considering just defining a native `nt._join_path` function that uses `PathCchCombineExW` and lets the OS worry about supporting all of that stuff. Too many edge cases for all the places we need this...

Reply:
> I'm considering just defining a native `nt._join_path`

Or implement `_path_abspath(p)` that calls `nt._getfullpathname(p)` in Windows, and `_path_join(os.getcwd(), p)` in POSIX.

For implementing `nt._join_path`, here's a few notable differences between `ntpath.join` and `PathCchCombineEx`.

The base API's core path functions, including `PathCchCombineEx`, only support backslash as the path separator. For example:

```python
import ctypes
corepath = ctypes.WinDLL('api-ms-win-core-path-l1-1-0', use_last_error=True)
p = (ctypes.c_wchar * 1000)()
```
right:
```python
>>> _ = corepath.PathCchCombineEx(p, len(p), "C:\\spam", "\\eggs", 0); p.value
'C:\\eggs'
```
wrong:
```python
>>> _ = corepath.PathCchCombineEx(p, len(p), "C:\\spam", "/eggs", 0); p.value
'C:\\spam\\/eggs'
```
I guess it's held over from the old path functions in shlwapi.dll, such as `PathCombineW`. In C, it's trivial to replace slashes with backslashes in a copy of each string. For internal use, the original path separators don't have to be preserved.

As documented, `PathCchCombineEx` canonicalizes a path to resolve "." and ".." components. In Windows, this only matters (i.e. may change how the path is resolved) with the target path of a relative symlink.  For example:

```python
>>> _ = corepath.PathCchCombineEx(p, len(p), "C:\spam", "foo\\..\\eggs", 0); p.value
'C:\\spam\\eggs'
```

`PathCchCombineEx` does not append drive-relative paths to the input path if they're relative to the same drive. For example:

```python
>>> _ = corepath.PathCchCombineEx(p, len(p), "C:\\spam", "C:eggs", 0); p.value
'C:eggs'
```

But no implementation of `_path_join(os.getcwd(), location)` can resolve a drive-relative path if it's not the same drive as the process working directory. The system uses "=X:" environment variables for this case, and defaults to the root path on the drive if the variable doesn't exist.

Reply:
> this only matters (i.e. may change how the path is resolved) with the target path of a relative symlink

Nope. I noticed that canonicalizing doesn't collapse repeated separators, so I decided to check how this interacts with resolving ".." components. The result is disappointing:

```python
>>> _ = corepath.PathCchCombineEx(p, len(p), "foo\\..\\\\spam", "eggs", 0); p.value
'\\\\spam\\eggs'
```
The type of path has been changed from a relative path to a UNC path.

Reply:
Yeah, I've been implementing this tonight (about to push changes), and it's an improvement, but not ideal. Unfortunately, there seems to be no way around a lot of code here to get consistency in importlib, and I'm happier with mostly following PathCchCombineEx behaviour than trying to invent our own.

Reply:
To be clear, the behaviour we'd be following is only in how it determines roots, so essentially the PathCchSkipRoot behaviour (to move splitdrive into native code). Only thing we lose is that the drive will always have normalised slashes, which seems to be okay to me.